src/mono/mono/component/debugger-agent.c

```diff
@@ -5416,11 +5432,184 @@ static ErrorCode decode_fixed_size_array_internal (MonoType *t, int type, MonoDo
 	*endbuf = buf;
 	return err;
 }
+
+static int
+decode_vtype_compute_size (MonoType *t, MonoDomain *domain, gpointer void_buf, guint8 **endbuf, guint8 *limit, gboolean from_by_ref_value_type)
+{
+	int ret = 0;
+	guint8 *buf = (guint8*)void_buf;
+	MonoClass *klass;
+	MonoClassField *f;
+	int nfields;
+	gpointer iter = NULL;
+	MonoDomain *d;
+	ErrorCode err;
+
+	/* is_enum, ignored */
+	decode_byte (buf, &buf, limit);
+	if (CHECK_PROTOCOL_VERSION(2, 61))
+		decode_byte (buf, &buf, limit);
+	klass = decode_typeid (buf, &buf, limit, &d, &err);
+	if (err != ERR_NONE)
+		goto end;
+
+	if (t && klass != mono_class_from_mono_type_internal (t)) {
+		goto end;
+	}
+
+	nfields = decode_int (buf, &buf, limit);
+	while ((f = mono_class_get_fields_internal (klass, &iter))) {
+		if (f->type->attrs & FIELD_ATTRIBUTE_STATIC)
+			continue;
+		if (mono_field_is_deleted (f))
+			continue;
+		if (!m_type_is_byref (f->type) && !m_class_is_byreflike (mono_class_from_mono_type_internal (f->type)) && !from_by_ref_value_type)
+			decode_value_compute_size (f->type, 0, domain, buf, &buf, limit, FALSE);
+		else
+			ret += decode_value_compute_size (f->type, 0, domain, buf, &buf, limit, m_type_is_byref (f->type) || m_class_is_byreflike (mono_class_from_mono_type_internal (f->type)));
```

Comment:
I was really confused what `from_by_ref_value_type` is used for.

But I think this is the only place where it actually makes a difference, right?

And its purpose is to decide if the current field needs extra space.

So we're saying - a value will need to be stored into extra space if:
1. It is byref, 
2. or it is a `ref struct`
3. or the current field is part of a value type that was byref in some outer struct.

It might be clearer to do something like:

```c
  /* byval members of the current field will need to be in extra_space */
  gboolean members_in_extra_space = m_type_is_byref (f->type) || m_class_is_byreflike (...);
  /* does the current field need to go in extra_space? Yes if it's recursively in a by_ref_value_type, or if it is itself byref or a ref struct */
  gboolean cur_field_in_extra_space = from_by_ref_value_type || members_in_extra_space;

  int field_size = decode_value_compute_size (f->type, 0, domain, buf, &buf, limit, members_in_extra_space);

  if (cur_field_in_extra_space)
    ret += field_size;
```


And actually I think I don't understand why it's like this.

Shouldn't it be somthing like:

```c
  gboolean cur_field_in_extra_space = from_by_ref_value_type;
  gboolean members_in_extra_space = cur_field_in_extra_space || m_type_is_byref (f->type);

  gsize field_size = decode_value_compute_size (f->type, ..., members_in_extra_space);
  if (cur_field_in_extra_space)
    ret += field_size;
```

ie once you're in extra_space, you're stuck in extra space for all the other structs you recurse into.  And additionallly if any of them have byref fields, those values need to be in extra_space too.

Reply:
Also if `f->type` is byref, don't you also need
  `if (cur_field_in_extra_space) ret += sizeof(gpointer);` ?

If the current field is `ref double` you need to store the double in `extra_space` and you also need to store a pointer in the current structure, I think.